_keystone/oidc-federation.md

```diff
@@ -0,0 +1,502 @@
+    scoped_token = Token(auth_url=auth_url,
+                         token=unscoped_auth_token._auth_token,
+                         project_id='<project-id>',
+                         reauthenticate=False)
+    federated_session = session.Session(auth=scoped_token)
+    nova = client.Client(version=2, session=federated_session)
+    for server in nova.servers.list():
+        print server.name
+
+if __name__ == '__main__':
+    main()
+```
+
+
+
+Using OAuth CLI for password flow
+
+```
+from keystoneauth1 import session
+from keystoneauth1.identity.v3 import oidc
+from keystoneauth1.identity.v3 import Token
+from keystoneclient.v3.client import Client
+
+from novaclient import client
+import os
+
+
+
+auth_url = os.environ.get('OS_AUTH_URL')
+identity_provider = os.environ.get('OS_IDENTITY_PROVIDER')
+protocol = 'oidc'
+redirect_uri = os.environ.get('OS_REDIRECT_URI')
+scope = ['openid','profile','email']
+client_id = os.environ.get('OS_CLIENT_ID')
+client_secret = os.environ.get('OS_CLIENT_SECRET')
+discovery_endpoint = os.environ.get('OS_DISCOVERY_ENDPOINT')
+access_token_endpoint = os.environ.get('OS_TOKEN_ENDPOINT')
+authorization_endpoint = os.environ.get('OS_AUTHORIZATION_ENDPOINT')
+
+def get_project_and_federated_session():
+    s = session.Session(verify=False)
+
+    # get an access token
+    payload = {'client_id': client_id, 'grant_type': 'password',
+                'username': '<username>',
+                'password': '<password>', 'scope': 'openid',
+                'connection':'Username-Password-Authentication'}
+
+    response = s.post('https://<FQDN-OP>/oauth/ro',
+                            data=payload,
+                            authenticated=False)
+    print(response.json())
+    access_token = response.json()['access_token']
+
+    oidc_plugin = oidc.OidcAccessToken(auth_url, identity_provider, protocol, access_token)
+    response = oidc_plugin.get_unscoped_auth_ref(s)
+    scoped_token = Token(auth_url=auth_url,
+                         token=response._auth_token,
+                         project_id='<project_id>',
+                         reauthenticate=False)
+    federated_session = session.Session(auth=scoped_token)
+    return  federated_session
+
+
+def main():
+    scoped_session = get_project_and_federated_session()
+    nova = client.Client(version=2, session=scoped_session)
+    for server in nova.servers.list():
+        print server.name
+
+if __name__ == '__main__':
+    main()
+```
+
+OIDC stackrc
+
+```
+export OS_AUTH_URL=https://<fqdn>:5000/v3
+export OS_IDENTITY_PROVIDER=<identity_provider name>
+export OS_PROTOCOL=<protocol_name>
+export OS_REDIRECT_URI=https://<fqdn>:5000/v3/auth/OS-FEDERATION/websso/oidc/redirect
+export OS_CLIENT_ID=<client_ID>
+export OS_CLIENT_SECRET=<client_secret>
+export OS_DISCOVERY_ENDPOINT=https://<FQDN-OP>/.well-known/openid-configuration
+export OS_TOKEN_ENDPOINT=https://<FQDN-OP>/oauth/token
+export OS_AUTHORIZATION_ENDPOINT=https://<FQDN-OP>/authorize
+```
```

Comment:
It would be nice to see some instruction to see how to run the script and source the credentials file.